### PR TITLE
add ESLint + rules

### PR DESCRIPTION
**Pull request by McBen**
ESLint and rules based on  https://github.com/IITC-CE/ingress-intel-total-conversion/wiki/Code-Style

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+  "name": "iitc-ce",
+  "version": "1.0.0",
+  "main": "core/total-conversion-build.js",
+  "repository": "https://github.com/IITC-CE/ingress-intel-total-conversion.git",
+  "author": ".",
+  "license": "MIT",
+  "dependencies": {
```
Comment (johnd0e) [inline package.json]:
```suggestion
  "devDependencies": {
```

File: package.json [hunk 2]
```diff
@@ -0,0 +1,13 @@
+{
+  "name": "iitc-ce",
+  "version": "1.0.0",
+  "main": "core/total-conversion-build.js",
+  "repository": "https://github.com/IITC-CE/ingress-intel-total-conversion.git",
+  "author": ".",
+  "license": "MIT",
```
Comment (johnd0e) [inline package.json]:
```suggestion
  "license": "ISC",
```

File: package.json [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  "dependencies": {
+    "eslint": "^6.8.0",
+    "eslint-plugin-es5": "^1.5.0",
+    "eslint-plugin-vue": "^6.2.0"
```
Comment (johnd0e) [inline package.json]: Do we really need these?

File: .eslintrc.json [hunk 4]
```diff
@@ -0,0 +1,38 @@
+{
+    "env": {
+        "browser": true,
+        "es6": false,
```
Comment (johnd0e) [inline .eslintrc.json:4]: It is `false` by default

File: .eslintrc.json [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        "SharedArrayBuffer": "readonly"
+    },
+    "parserOptions": {
+        "ecmaVersion": 2018,
+        "sourceType": "module"
+    },
+    "plugins": [
+        "vue",
+        "es5"
+    ],
```
Comment (johnd0e) [inline .eslintrc.json]: Do we really need these?

File: .eslintrc.json [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        "eslint:recommended",
+        "plugin:vue/essential",
+        "plugin:es5/no-es2016"
+    ],
+    "globals": {
+        "Atomics": "readonly",
+        "SharedArrayBuffer": "readonly"
+    },
+    "parserOptions": {
+        "ecmaVersion": 2018,
```
Comment (johnd0e) [inline .eslintrc.json]: Why not just `"ecmaVersion": 5`

File: .eslintrc.json [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        "greasemonkey": true
+    },
+    "extends": [
+        "eslint:recommended",
+        "plugin:vue/essential",
+        "plugin:es5/no-es2016"
+    ],
+    "globals": {
+        "Atomics": "readonly",
+        "SharedArrayBuffer": "readonly"
```
Comment (johnd0e) [inline .eslintrc.json]: What are these?

File: .eslintrc.json [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    },
+    "parserOptions": {
+        "ecmaVersion": 2018,
+        "sourceType": "module"
```
Comment (johnd0e) [inline .eslintrc.json]: Our sources are not ECMAScript modules

File: .eslintrc.json [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    "env": {
+        "browser": true,
+        "es6": false,
+        "jquery": true,
+        "greasemonkey": true
+    },
+    "extends": [
+        "eslint:recommended",
+        "plugin:vue/essential",
+        "plugin:es5/no-es2016"
```
Comment (johnd0e) [inline .eslintrc.json]: Do we really need these?

File: package.json [hunk 10]
```diff
@@ -0,0 +1,13 @@
+{
+  "name": "iitc-ce",
+  "version": "1.0.0",
```
Comment (johnd0e) [inline package.json]:
```suggestion
  "version": "0.1.0",
```

File: package.json [hunk 11]
```diff
@@ -0,0 +1,13 @@
+{
+  "name": "iitc-ce",
```
Comment (johnd0e) [inline package.json:2]: Maybe also
```
  "description": "ingress intel total conversion (IITC-CE)",
```
```
  "keywords": [
    "iitc",
    "ingress"
  ],
```
```
  "private": true,
```

File: .eslintrc.json [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+    "parserOptions": {
+        "ecmaVersion": 2018,
+        "sourceType": "module"
+    },
+    "plugins": [
+        "vue",
+        "es5"
+    ],
+    "rules": {
+        "no-undef": 0,
```
Comment (johnd0e) [inline .eslintrc.json]:
```suggestion
        "no-undef": "error"
```
Bot it is default value, so it is not necessary to specify it explicitly.

File: .eslintrc.json [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+        "eqeqeq": "error",
+        "indent": ["warn", 2],
+        "brace-style": "error",
```
Comment (johnd0e) [inline .eslintrc.json]:
```suggestion
        "brace-style": ["error","1tbs", { "allowSingleLine": true }],
```

File: .eslintrc.json [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+        "keyword-spacing": ["error", { "before": true }],
```
Comment (johnd0e) [inline .eslintrc.json]:
```suggestion
        "keyword-spacing": "error",
```
No need in `before` as it by default.

File: .eslintrc.json [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+        "spaced-comment": ["error", "always"],
```
Comment (johnd0e) [inline .eslintrc.json]:
```suggestion
        "spaced-comment": "error",
```

File: .eslintrc.json [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+        "quotes": ["error", "single", { "avoidEscape": true }],
+        "no-const-assign": "error",
```
Comment (johnd0e) [inline .eslintrc.json]: No need in es5, as it is anyway  handled like this:
```
Parsing error: The keyword 'const' is reserved [Error]
```

File: .eslintrc.json [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+        "no-trailing-spaces": "error",
+        "max-len": ["warn", { "code": 160 }]
+        // "no-var": "error"
```
Comment (johnd0e) [inline .eslintrc.json]:
```suggestion
```

**Review by johnd0e — CHANGES_REQUESTED** [review]
And more:
- add https://eslint.org/docs/rules/semi
- remove `yarn.lock`

**Comment by johnd0e** [comment]
Todo: may be add more rules, like:
- [eol-last](https://eslint.org/docs/rules/eol-last)
- [quote-props](https://eslint.org/docs/rules/quote-props) (`"quote-props": ["error","consistent-as-needed"],`)
- [no-unused-expressions](https://eslint.org/docs/rules/no-unused-expressions)
- etc

File: package.json (hunk 2 above)
Comment (McBen) [reply package.json]: changed

File: package.json (hunk 3 above)
Comment (McBen) [reply package.json]: removed

File: package.json (hunk 1 above)
Comment (McBen) [reply package.json]: fixed

File: .eslintrc.json (hunk 4 above)
Comment (McBen) [reply .eslintrc.json:4]: it's just to express we are using es5

File: .eslintrc.json (hunk 5 above)
Comment (McBen) [reply .eslintrc.json]: removed

File: .eslintrc.json (hunk 13 above)
Comment (McBen) [reply .eslintrc.json]: add

File: .eslintrc.json (hunk 7 above)
Comment (McBen) [reply .eslintrc.json]: note sure. were added by eslint --init. I guess related to vue?

File: .eslintrc.json (hunk 9 above)
Comment (McBen) [reply .eslintrc.json]: removed

File: .eslintrc.json (hunk 14 above)
Comment (McBen) [reply .eslintrc.json]: done

File: .eslintrc.json (hunk 15 above)
Comment (McBen) [reply .eslintrc.json]: done

File: .eslintrc.json (hunk 6 above)
Comment (McBen) [reply .eslintrc.json]: see below

File: .eslintrc.json (hunk 8 above)
Comment (McBen) [reply .eslintrc.json]: removed

File: package.json (hunk 10 above)
Comment (McBen) [reply package.json]: removed version - unused anyway

File: package.json (hunk 11 above)
Comment (McBen) [reply package.json:2]: what ever you wish ... unused anyway

File: .eslintrc.json (hunk 16 above)
Comment (McBen) [reply .eslintrc.json]: removed

File: .eslintrc.json (hunk 12 above)
Comment (McBen) [reply .eslintrc.json]: removed

File: .eslintrc.json (hunk 7 above)
Comment (McBen) [reply .eslintrc.json]: removed

File: .gitignore [hunk 18]
```diff
@@ -8,6 +8,8 @@ local*.py
 .pydevproject
 .DS_Store
 
+node_modules
+
```
Comment (johnd0e) [inline .gitignore:12]: Shouldn't be part of this PR

File: .eslintrc.json [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+        "spaced-comment": "error",
+        "quotes": ["error", "single", { "avoidEscape": true }],
+        "no-trailing-spaces": "error",
+        "max-len": ["warn", { "code": 160 }],
+        "semi": "error",
+        "eol-last": "error",
+        "quote-props": ["error","consistent-as-needed"],
+        "no-unused-expressions": "error"
+    }
+}
```
Comment (johnd0e) [inline .eslintrc.json]: Perhaps EOL needed here

File: package.json [hunk 20]
```diff
@@ -0,0 +1,16 @@
...
+    "iitc",
+    "ingress"
+  ],
+  "main": "core/total-conversion-build.js",
+  "repository": "https://github.com/IITC-CE/ingress-intel-total-conversion.git",
+  "author": ".",
+  "license": "ISC",
+  "private": true,
+  "devDependencies": {
+    "eslint": "^7.5.0"
```
Comment (johnd0e) [inline package.json]: Not sure about it, because in fact it is not used yet.
And current version is 7.16, so may be better just keep `^7` (no difference, but more expessive)

File: .eslintrc.json [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+    "rules": {
+        "eqeqeq": "error",
+        "indent": ["warn", 2],
+        "brace-style": ["error","1tbs", { "allowSingleLine": true }],
+        "keyword-spacing": "error",
+        "spaced-comment": "error",
+        "quotes": ["error", "single", { "avoidEscape": true }],
+        "no-trailing-spaces": "error",
+        "max-len": ["warn", { "code": 160 }],
+        "semi": "error",
```
Comment (johnd0e) [inline .eslintrc.json:25]: It would be more convenient to have the same rules order as we have is style guide: https://github.com/IITC-CE/ingress-intel-total-conversion/wiki/Code-Style

**Review by johnd0e — CHANGES_REQUESTED** [review]

**Comment by johnd0e** [comment]
@modos189 
Please review these rules: https://github.com/IITC-CE/ingress-intel-total-conversion/pull/335#issuecomment-588163570
They weren't part of original style guide (but we include if you are ok with them).

Alternatively, we may set them as warnings (not errors).

File: .gitignore (hunk 18 above)
Comment (McBen) [reply .gitignore:12]: This PR introduces eslint node_package.
Adding packages without this would be a really bad idea.

Comment (johnd0e) [reply .gitignore:12]: Ah, right.

File: package.json (hunk 20 above)
Comment (McBen) [reply package.json]: could be updated, yes - when doing this PR 7.5 was latest-release. 

I don't like automatic version updates of underlying libs. (eslint might be an exception here - but I don't like exception either :) )

File: .eslintrc.json (hunk 19 above)
Comment (McBen) [reply .eslintrc.json]: not required but could be added

File: .eslintrc.json (hunk 21 above)
Comment (McBen) [reply .eslintrc.json:25]: it's a configuration file - noone will every 'read' these files. There is a loosly connetion to the wikipage not a dependency.

**Comment by McBen** [comment]
Maybe the rule `no-undef` and `no-unused-vars` should be disabled. 

They will generate false-positives because the custom build-script concept isn't really supported by any other tool.

**Comment by johnd0e** [comment]
> Maybe the rule `no-undef` and `no-unused-vars` should be disabled.

Do you mean 'enabled'? Because you do not have those in the PR.

> They will generate false-positives because the custom build-script concept isn't really supported by any other tool.

Explain this please. I am not sure what false positives do you mean, and how they are related to some custom build-script.

**Comment by McBen** [comment]
> > Maybe the rule `no-undef` and `no-unused-vars` should be disabled.
> 
> Do you mean 'enabled'? Because you do not have those in the PR.

This rule-set is based on `eslint:recommended`

 
> > They will generate false-positives because the custom build-script concept isn't really supported by any other tool.
> 
> Explain this please. I am not sure what false positives do you mean, and how they are related to some custom build-script.

Examples:
```
./core/total-conversion-build.js
    8:22  error    'plugin_info' is not defined                                           no-undef
   23:24  error    'PLAYER' is not defined                                                no-undef
  177:5   error    'portalsFactionLayers' is defined but never used                       no-unused-vars
  177:27  error    'linksFactionLayers' is defined but never used                         no-unused-vars
  177:47  error    'fieldsFactionLayers' is defined but never used                        no-unused-vars
```

**Comment by johnd0e** [comment]
I see now. But I do not agree: those are not false-positives, those are places for improvements, as we definitely are able to clearly express our intentions, in number of ways, e.g.:
- use `window.` notation
- use exclusion list (custom list of globals)
- use inline exclusions (in comments)

~~And I still do not understand how this is related to custom build script.
We have no linting step now, true. But it is easy to add (when the sources will be ready).~~

**Update**
Perhaps you mean that in webpack there will be explicit import/export statements.
BTW:
```js
var portalsFactionLayers = [];
// ...........................
window.portalsFactionLayers = portalsFactionLayers; // example of some primitive form of export
```
Then:
```js
var portalsFactionLayers = window.portalsFactionLayers;  // example of some primitive form of import
```

**Comment by johnd0e** [comment]
https://eslint.org/docs/rules/curly#all
?

File: package.json [hunk 22]
```diff
@@ -0,0 +1,16 @@
...
+    "iitc",
+    "ingress"
+  ],
+  "main": "core/total-conversion-build.js",
+  "repository": "https://github.com/IITC-CE/ingress-intel-total-conversion.git",
+  "author": ".",
+  "license": "ISC",
+  "private": true,
+  "devDependencies": {
+    "eslint": "^7.16.0"
```
Comment (johnd0e) [inline package.json]: https://eslint.org/blog/2021/02/eslint-v7.20.0-released

Comment (McBen) [reply package.json]: Thats the job of the package manager. There is no need to keep track of every update.
`npm install`/`yarn install` (or upgrade) will handle it by themself.

**Comment by McBen** [comment]
> 
> 
> https://eslint.org/docs/rules/curly#all
> ?


d9f4e7f

**Comment by johnd0e** [comment]
@McBen 
Do you insist on `multi-line` option?
I'd prefer `all`.

**Comment by McBen** [comment]
> Do you insist on `multi` option?
> I'd prefer `all`.

I don't insist on anything. I just like to use constructs like 
`if (error) return;` 
(instead of `if (error) { return; }`)

**Comment by modos189** [comment]
I like the following options:

`eol-last: always`
`quote-props: always`
`curly: multi-line`

It hardly makes sense to use `no-undef` and `no-unused-vars`, because we are using our own build system instead of webpack and this should confuse eslint.

But what about third-party libraries? Possible incompatibilities with their style won't cause problems?

**Comment by McBen** [comment]
> `eol-last: always`

is default

> `quote-props: always`

code like this would trigger an error:
`marker.setStyle({ opacity: 0.4, fillOpacity: 0.1 });`

> `curly: multi-line`

makes 2 votes for multi-line 
 
> But what about third-party libraries? Possible incompatibilities with their style won't cause problems?

commit 55066a0 - ESLint: Ignore externals
but since we never edit external source this should not be a prolbem ;)

**Comment by johnd0e** [comment]
> > `eol-last: always`
> 
> is default

Where do you see such defaults?
Because it is not check-marked in [rules list](https://eslint.org/docs/rules/).

**Comment by McBen** [comment]
> 
> 
> > > `eol-last: always`
> > 
> > 
> > is default
> 
> Where do you see such defaults?
> Because it is not check-marked in [rules list](https://eslint.org/docs/rules/).

https://eslint.org/docs/rules/eol-last

eol-last is not a default rule. But 'always' is the default option for eol-last.

File: package.json [hunk 23]
```diff
@@ -0,0 +1,16 @@
+{
+  "name": "iitc-ce",
+  "description": "ingress intel total conversion (IITC-CE)",
+  "keywords": [
+    "iitc",
+    "ingress"
+  ],
+  "main": "core/total-conversion-build.js",
```
Comment (johnd0e) [inline package.json]: I'm not sure about this field.
E.g. I see that Leaflet uses built script path for it: https://github.com/Leaflet/Leaflet/blob/234e00bcab0d89319f4c6874f3063ac311ce8cc6/package.json#L33

File: package.json [hunk 24]
```diff
@@ -0,0 +1,16 @@
...
+  "repository": "https://github.com/IITC-CE/ingress-intel-total-conversion.git",
+  "author": ".",
```
Comment (johnd0e) [inline package.json]:
```suggestion
```
What's meant here?

File: package.json (hunk 23 above)
Comment (le-jeu) [reply package.json]: same as `author`, is this usefull ? this is not a package (yet?)

**Review by johnd0e — APPROVED** [review]

**Review by modos189 — APPROVED** [review]
